features/370-auth-ui/design.md

```diff
@@ -0,0 +1,393 @@
+
+The value must be `code` because the implementation only supports Authorization Code Flow.
+
+#### client_id
+
+The value must be a valid Skygear API Key.
+
+#### redirect_uri
+
+No difference from the spec.
+
+#### state
+
+It is ignored.
+
+#### nonce
+
+It is ignored.
+
+#### display
+
+It is not supported. The only supported display mode is `page`. Note that this parameter corresponds to UXMode in the SSO feature. In SSO, UXMode can be `redirect` or `popup`, which are equivalent to `page` and `popup` in the OIDC Core spec.
+
+#### prompt
+
+It is not supported.
+
+#### max_age
+
+It is not supported.
+
+#### ui_locales
+
+No difference from the spec.
+
+#### id_token_hint
+
+It is not supported.
+
+#### login_hint
+
+It is not supported.
+
+#### acr_values
+
+It is not supported.
+
+---
+
+#### code_challenge_method
+
+Only `S256` is supported. `plain` is not supported. The reason is the existing implementation of SSO is using `S256` already. Since it defaults to `plain`, it is required and the value must be `S256`.
+
+#### code_challenge
+
+No difference from the spec.
+
+#### Example of valid Authorization Request
+
+`response_type=code&scope=openid&client_id=API_KEY&redirect_uri=REDIRECT_URI&code_challenge_method=S256&code_challenge=CODE_CHALLENGE&ui_locales=en`
+
+### Token Endpoint
+
+The following sessions list out the request parameters and their difference from the spec.
+
+#### grant_type
+
+The value must be `authorization_code` because the implementation only supports Authorization Code Flow.
+
+#### code
+
+No difference from the spec.
+
+#### redirect_uri
+
+No difference from the spec.
+
+#### client_id
+
+It is always required. The value must be a valid Skygear API Key.
+
+#### code_verifier
+
+No difference from the spec.
+
+#### Example of valid Token Request
+
+`grant_type=authorization_code&code=AUTHORIZATION_CODE&redirect_uri=REDIRECT_URI&client_id=API_KEY&code_verifier=CODE_VERIFIER`
+
+### Token Response
+
+Since we have to support both cookie and header transport, the difference here is very large.
+
+#### id_token
+
+It is always absent. It is the JWT token that we may want to implement in the future.
+
+#### token_type
+
+If the API key is header transport, its value is `bearer`. Otherwise this key is absent.
+
+#### access_token
+
+If the API key is header transport, its value is the Skygear access token. Otherwise it is absent and the access token is set in HTTP cookie header instead.
+
+#### expires_in
+
+It is always absent.
+
+#### refresh_token
+
+If the API key is header transport, its value is the Skygear refresh token. Otherwise it is absent.
+
+#### scope
+
+It is always absent.
+
+#### Example of Token Response with header transport
+
+```json
+{
+  "token_type": "bearer",
+  "access_token": "access.token",
+  "refresh_token": "refresh.token"
+}
+```
+
+#### Example of Token Response with cookie transport
+
+```json
+{}
+```
+
+## Security considerations
+
+### Being embedded in an HTML iframe
```

Comment:
I'm not sure if safari's anti tracking cookie feature would affect this. We may need to look into this.

Reply:
Just tested with the following program

```golang
package main

import (
	"fmt"
	"math/rand"
	"net/http"
	"strconv"
)

func main() {
	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
		value := strconv.Itoa(rand.Int())
		cookie := &http.Cookie{
			Name:     "foobar",
			Value:    value,
			Path:     "/",
			HttpOnly: true,
		}
		body := fmt.Sprintf(`
<!DOCTYPE html>
<html>
<head>
<title>/</title>
<body>
/ %s
<iframe src="http://iframe.testsafari.localhost:9999/iframe" />
</body>
</head>
</html>
		`, value)
		http.SetCookie(w, cookie)
		_, _ = w.Write([]byte(body))
	})
	http.HandleFunc("/iframe", func(w http.ResponseWriter, r *http.Request) {
		cookie, err := r.Cookie("foobar")
		value := "<none>"
		if err == nil {
			value = cookie.Value
		}
		body := fmt.Sprintf(`
<!DOCTYPE html>
<html>
<head>
<title>/</title>
<body>
iframe %s
</body>
</head>
</html>
		`, value)
		_, _ = w.Write([]byte(body))
	})
	_ = http.ListenAndServe(":9999", nil)
}

```

Open Safari and go to `http://root.testsafari.localhost:9999`. The iframe cannot read the cookie. However, if the origin is the same (`http://iframe.testsafari.localhost:9999`), cookie can be read in the iframe. I guess as long as we keep the origin being the same, the cookie is considered first party and can be read without problems.